### PR TITLE
dialects: (memref) add custom syntax for memref_subview

### PR DESCRIPTION
**Pull request by superlopuh**
I tried to match MLIR exactly in the syntax, and found a bug in the stencil lowering, which lowers to an invalid memref subview. My understanding is that the subview must be a subview, and so negative offsets are not allowed.

I won't merge without someone from the stencil side of things approving the PR, but also I'd love to use this for the linalg to snitch project, so I'm hoping that we can resolve this soonish.

### CODE REVIEW

File: tests/filecheck/transforms/convert-stencil-to-ll-mlir.mlir [hunk 1]
```diff
@@ -292,51 +292,6 @@ builtin.module {
...
-      %58 = stencil.access %57[1] : !stencil.temp<[1,65]xf64>
-      stencil.return %58 : f64
-    }
-    stencil.store %56 to %50 ([0] : [64]) : !stencil.temp<[0,64]xf64> to !stencil.field<[-4,68]xf64>
-    func.return
-  }
-
-// CHECK:         func.func @stencil_buffer(%0 : memref<72xf64>, %1 : memref<72xf64>) {
-// CHECK-NEXT:      %2 = memref.alloc() : memref<64xf64>
-// CHECK-NEXT:      %3 = "memref.subview"(%2) <{"static_offsets" = array<i64: -1>, "static_sizes" = array<i64: 64>, "static_strides" = array<i64: 1>, "operandSegmentSizes" = array<i32: 1, 0, 0, 0>}> : (memref<64xf64>) -> memref<64xf64, strided<[1], offset: -1>>
```
Comment (superlopuh) [inline tests/filecheck/transforms/convert-stencil-to-ll-mlir.mlir:312]: Negative offset, not allowed.

- [ ] need to add this back

File: tests/filecheck/transforms/convert-stencil-to-ll-mlir.mlir [hunk 2]
```diff
@@ -713,74 +668,6 @@ func.func @if_lowering(%arg0_1 : f64, %b0 : !stencil.field<[0,7]x[0,7]x[0,7]xf64
...
-      %127 = arith.addf %125, %126 : f64
-      stencil.return %127 : f64
-    }
-    stencil.store %123 to %116 ([1, 2] : [65, 66]) : !stencil.temp<[1,65]x[2,66]xf64> to !stencil.field<[-3,67]x[-3,67]xf64>
-    func.return
-  }
-
-// CHECK:         func.func @buffered_combine(%0 : memref<?x?xf64>) {
-// CHECK-NEXT:      %1 = memref.alloc() : memref<64x64xf64>
-// CHECK-NEXT:      %2 = "memref.subview"(%1) <{"static_offsets" = array<i64: -1, -2>, "static_sizes" = array<i64: 64, 64>, "static_strides" = array<i64: 1, 1>, "operandSegmentSizes" = array<i32: 1, 0, 0, 0>}> : (memref<64x64xf64>) -> memref<64x64xf64, strided<[64, 1], offset: -66>>
```
Comment (superlopuh) [inline tests/filecheck/transforms/convert-stencil-to-ll-mlir.mlir:740]: Negative offset, not allowed.

- [ ] need to add this back

File: xdsl/dialects/memref.py [hunk 3]
```diff
@@ -591,6 +595,9 @@ def get_canonicalization_patterns(cls) -> tuple[RewritePattern, ...]:
 
 @irdl_op_definition
 class Subview(IRDLOperation):
+
+    DYNAMIC_INDEX: ClassVar[int] = -9223372036854775808
```
Comment (webmiche) [inline xdsl/dialects/memref.py:599]: Obviously

Comment (superlopuh) [reply xdsl/dialects/memref.py:599]: 🔢

**Review by PapyChacal — APPROVED** [review]
Lovely, thanks!

I would be extra glad if you could get rid of the extra space between space brackets? But if it's more than 5 minutes of thinking, please just merge, it's obviously 100X better already :pray:

**Comment by superlopuh** [comment]
So I initially did it without the space, and then added it back in to be 1:1 conformant with MLIR syntax.

File: tests/filecheck/mlir-conversion/with-mlir/dialects/memref/memref_ops_custom.mlir [hunk 4]
```diff
@@ -46,4 +48,7 @@ memref.dealloc %a1 : memref<2x3xf32>
 // CHECK-NEXT:   %{{.*}} = memref.alloc() : memref<2x3xf32>
 // CHECK-NEXT:   %{{.*}} = memref.alloc()[%{{.*}}] {alignment = 8 : i64} : memref<2x3xf32, affine_map<(d0, d1)[s0] -> (d0 + s0, d1)>, 1>
 // CHECK-NEXT:   memref.dealloc %{{.*}} : memref<2x3xf32>
+// CHECK-NEXT:   %{{.*}} = memref.subview %{{.*}}[0, 0] [1, 3] [1, 1] : memref<10x3xi32> to memref<3xi32>
+// CHECK-NEXT:   %{{.*}} = memref.subview %{{.*}}[%{{.*}}, 0] [1, 3] [1, 1] : memref<10x3xi32> to memref<3xi32, strided<[1], offset: ?>>
```
Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/memref/memref_ops_custom.mlir:52]: @PapyChacal, this is the output from MLIR. It looks like the doc online disagrees, but I'd rather go with the MLIR we have as source of truth.

**Pull request merged**